Add alternate autofill option for games other than Breath of the Wild

itsschwer opened the pull request:
The event editor is compatible with Tears of the Kingdom but the the file structure has changed, meaning that the autofill feature is unusable for it. I've tried to add an alternative option that requires more manual work on the user's part but allows them to setup autofill for Tears of the Kingdom (and potentially other games?). The readme has also been updated with a brief usage guide.

Additionally, this is my first time working with Python, so additional changes may be required (will gladly accept any feedback!)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zeldamods/event-editor/8)
<!-- Reviewable:end -->

itsschwer committed:
Add autofill fallback using manual json

itsschwer committed:
Update README.md

itsschwer committed:
readme: fix typo

leoetlino commented:
IMO the format should be a single file containing action/query definitions for all actors. Actions and queries should also be separate, and it should be clearer what the values are used for.  I'd suggest something like this:

```json
{
  "Actor1": {
    "actions": {
      "Action1": {
        "params": {
          "Param1": {"default_value": true},
          "Param2": {"default_value": 42},
        }
      }
    },
    "queries": {} // same format as actions
  },
}
```

Using JSON for this sounds fine since this kind of file is meant to be automatically generated.

Also, I would call this file an "actor definition file" - "JSON dir" sounds a bit too vague/generic.

itsschwer committed:
Change json loading to expect a file per actor

itsschwer commented:
To clarify the use case, the JSON files are (currently) expected to be user-generated, so that they can add their own parameters for use with autofill. This is why I've opted to use a more minimal/flatter JSON structure, though I do think having a file per action/query is excessive.

As for action and query separation, I'm actually not super familiar with the event flow structure. Is it just that actions belong to Action events and queries belong to Switch events?

leoetlino commented:
Yes, and actions/queries can theoretically have the same name so they should be separate.

The reason why I suggested a more complex schema is to make it future-proof (easily extendable) and to make it easier for someone to efficiently auto-generate an EventEditor compatible actor definition file for games like TotK. Creating one file per actor is not super convenient when you have 7000+ actors.

And ideally, we'd also replace the BotW AI specific logic with a separate tool that generates the JSON definitions, so that the JSON file is not just an afterthought/fallback but a flexible mechanism of adding autocomplete (and maybe action/query documentation) - one that works with any game.

itsschwer committed:
Create actor_json.py and update JSON schema

- Move JSON handling into separate file
- Separate "actions" and "queries" in JSON schema

itsschwer committed:
Extend json support to action/query additions

itsschwer committed:
Export actors as json (event names only)

itsschwer committed:
Use actor name for exported json file name

itsschwer commented:
I've changed the pull request status to draft as I've started to add the export/JSON generation functionality but it's incomplete.

I see your point about using a more complex and descriptive schema for the JSON definitions, but at this stage (and as a user) I wish to continue with a one actor one file approach (so far I haven't encountered a large number of different event actors?) — I'm open to changing this in the future though! Currently the schema looks like this:
```jsonc
// Actor1.json
{
    "actions": {
        "Action1": {
            "Parameter1": true
        }
    },
    "queries": {
        // etc
    }
}
```
Also, I'm currently a bit stumped on how to populate the parameters with default/sample values, would you be able to provide some direction?

Thanks!

itsschwer committed:
Add 'Copy JSON' button for event parameters

itsschwer committed:
Fix typo

itsschwer committed:
actor_json: clean up type hints

itsschwer committed:
readme: update auto-completion instructions

itsschwer committed:
actor_json: use QWidget type hint

leoetlino commented:
> I see your point about using a more complex and descriptive schema for the JSON definitions, but at this stage (and as a user) I wish to continue with a one actor one file approach (so far I haven't encountered a large number of different event actors?) — I'm open to changing this in the future though!

I think we should avoid adding a feature only to remove or drastically change how it works later.

In my opinion the UX for the mechanism that's being added in this PR is not great -- it shouldn't be the case that the user has to mess with a JSON file every single time they want to add an action or query to the actor definition set. 

Instead, there should be a way to batch export definitions for all actions/queries that are used in an event flow.

> Also, I'm currently a bit stumped on how to populate the parameters with default/sample values, would you be able to provide some direction?

This is easy to implement with the batch export idea. Just iterate through the actions and switches in the event flowchart; that way you get both the action/query names and their parameters at the same time.

itsschwer committed:
actor_json: Export to single combined file

itsschwer committed:
actor_json: Read from single combined file

itsschwer committed:
Remove json fallback error message

Inaccurate, completely replaced ai error rather than only when json exists but missing parameters

itsschwer committed:
actor_json: Add todo comments

itsschwer commented:
I see where you're coming from about the complexity for general users — I've altered the code to use a single JSON file (but prefer a specific actor file if it exists as an 'override' when loading).

The UX for exporting is still very much a work-in-progress, hence the draft status. The current implementation is just to get something working as I've also been using the tool (and people in the Tears of the Kingdom modding server have found my fork).

The batch exporting is definitely a goal for this pull request, but I'm not experienced enough to know where to start. Would you be able to provide more detail about where would be a good entry point to get the data for iterating?

leoetlino commented:
There is an "Export graph data to JSON" feature (for exporting the entire flowchart, not just action/query info) that you can mimic.

https://github.com/zeldamods/event-editor/blob/a30f5db70a7c8ac948c35b1a37638a8768555470/eventeditor/__main__.py#L104

https://github.com/zeldamods/event-editor/blob/a30f5db70a7c8ac948c35b1a37638a8768555470/eventeditor/flowchart_view.py#L193

`self.flow_data.flow` (in FlowchartView) is the `EventFlow` object for the currently loaded event flow: https://github.com/zeldamods/evfl/blob/645dc7d786233896561ea27a697e7a171c0828cb/evfl/evfl.py#L10

`flow.flowchart.events` (https://github.com/zeldamods/evfl/blob/645dc7d786233896561ea27a697e7a171c0828cb/evfl/flowchart.py#L9) is a list of flowchart events that you can iterate on. Take a look at `evfl/event.py` to see what kind of fields an Event has (I've never written detailed docs for evfl, sorry! but hopefully the source code is easy enough to read)

---

edit: some additional context. `evfl` is the Python library that deals with event flow parsing/serialisation. It's what EventEditor (the graphical editor) is built upon.

itsschwer committed:
Replace individual actor export with batch export

Produces same output, no event parameters yet

itsschwer committed:
actor_json: Fully implement export_definitions

itsschwer committed:
event_edit_dialog: Add 'Paste JSON' button

itsschwer committed:
event_edit_dialog: Add paste fail error message

itsschwer committed:
Manually set save shortcut

QKeySequence standard's 'SaveAs' does not support Windows

itsschwer committed:
Update .gitignore

itsschwer committed:
readme: Update json auto-completion section

itsschwer commented:
Thanks for the links, they were just what I needed!

I think my implementation is ready for review now, though I'm not sure I've gone about it in a clean way. I'm not sure if this is good practice, but I've left comments to explain my thought process or in areas where I was uncertain — all of this should be in `actor_json.py`.

To summarise the feature changes:
- Add ability to copy event parameters using JSON
- Add ability to paste event parameters using JSON
- Add ability to export JSON definitions for an open event flow
- Auto-completion now looks for JSON definitions when ROM method fails (event parameters; adding actor actions/queries)

Thanks again for the guidance!

leoetlino commented on eventeditor/__main__.py:
```diff
@@ -180,6 +186,7 @@ def closeEvent(self, event) -> None:
     def readSettings(self) -> None:
         settings = qc.QSettings()
         ai.set_rom_path(settings.value('paths/rom_root'))
+        aj.set_actor_json_path(settings.value('paths/actor_json_root'))
```
nit: I would recommend calling this actor_definitions as just "actor" is a bit vague. JSON shouldn't be in the name as we don't want to be tied to a specific storage format.

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+    except:
+        return None
+
+def load_actions(actor_name: str) -> typing.KeysView[str]:
```
`Iterable[str]` (callers don't need to know that it's a KeysView, just that it's something that can be iterated on and that produces strings)

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+        return
+
+    if not _actor_json_path:
+        set_actor_json_path(q.QFileDialog.getSaveFileName(widget, 'Set ',  'actor_definitions', 'JSON (*.json)')[0])
```
"Set "?

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+    except:
+        return None
+
+def export_definitions(flow, widget) -> None:
```
nit: please add type annotations here for flow and widget

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+        definitions = dict()
+    
+    for actor in flow.flowchart.actors:
+        if actor.identifier.name not in definitions:
```
```py
actor_def = definitions.get(actor.identifier.name, {})
```
and then use actor_def instead of repeating definitions[actor.identifier.name] all the same

and at the end:
```py
definitions[actor.identifier.name] = actor_def
```

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+        if actor.identifier.name not in definitions:
+            definitions[actor.identifier.name] = {}
+
+        if 'actions' not in definitions[actor.identifier.name]:
```
ditto, use .get() and [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+            if query.v not in definitions[actor.identifier.name]['queries']:
+                definitions[actor.identifier.name]['queries'][query.v] = {}
+        
+        #! Should actor parameters also be supported?
```
Feel free to leave this for a future PR (or not).

itsschwer replied:
I'll leave this for a potential future PR since it seems like a new (separate) feature to event-editor

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+    
+    #! Should exporting be restructured to only loop through the events?
+    for event in flow.flowchart.events:
+        if hasattr(event.data, 'actor_action'):
```
Check the event types instead of checking for the existence of specific property names (which is more fragile, doesn't really work with mypy and will break silently if we ever rename the actor_action field -- the worst kind of breakage).

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,121 @@
+            continue
+
+        if event_key not in json_parent:
+            # Shouldn't be reachable, as event names are populated in actor loop?
```
This seems fine to have as a "just in case" thing.

leoetlino commented on eventeditor/flowchart_view.py:
```diff
@@ -202,6 +203,12 @@ def export(self) -> None:
+        try:
+            aj.export_definitions(self.flow_data.flow, self)
+        except:
+            q.QMessageBox.critical(self, 'Export actor definition data', 'Failed to write to ' + path)
```
path is not defined anywhere? Make sure you run a linter and/or mypy

itsschwer committed:
Fix autofill typo and export error message

itsschwer committed:
Rename 'actor_json' to 'actor_definitions'

itsschwer committed:
actor_json: Check event type instead of attributes

itsschwer committed:
actor_json: Update type hints

itsschwer committed:
actor_json: Refactor export_definitions

itsschwer commented:
Thanks for the detailed review!

I've tried to address all of the sections you've highlighted, but I'm not completely sure if I've done the type hinting right.

*re: linter/mypy*
I'm not very familiar with the Python workflow, my bad for letting those errors slip through! I'm working in VS Code and have now installed a [Mypy extension](https://marketplace.visualstudio.com/items?itemName=matangover.mypy), but it seems to identify a large number of errors (304) across the project — would you happen to have any experience with this? (all good if not!)

itsschwer committed:
actor_json: fix minor style inconsistency

Order shouldn't matter as Python always passes by reference(?)

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,119 @@
+    try:
+        # First look for individual actor file (overrides; for power users)
+        with open(_actor_definitions_path.parent/f'{actor_name}.json', 'rt') as file:
+            return json.loads(file.read())
```
json.load(file)

leoetlino commented on eventeditor/actor_json.py:
```diff
@@ -0,0 +1,119 @@
+        try:
+            # Otherwise look in actor definitions file
+            with open(_actor_definitions_path, 'rt') as file:
+                return json.loads(file.read())[actor_name]
```
json.load(file)

leoetlino reviewed (approved):
lgtm, just one minor thing

leoetlino commented:
> I'm not very familiar with the Python workflow, my bad for letting those errors slip through! I'm working in VS Code and have now installed a [Mypy extension](https://marketplace.visualstudio.com/items?itemName=matangover.mypy), but it seems to identify a large number of errors (304) across the project — would you happen to have any experience with this? (all good if not!)

It's been a while since I did any major development work on evfl or event-editor, but iirc PyQt sometimes causes mypy to get really confused. If the errors you're seeing are about PyQt, I wouldn't worry too much about them. We might be able to make them go away by updating our dependencies.

itsschwer committed:
actor_json: Replace json.loads() with json.load()
